File: Compiler/src/compiler/operators/CheckRelationship.java
package compiler.operators;

import compiler.Operator;
import compiler.values.DecisionTreeVarValue;
import compiler.values.RelationshipValue;
import dictionaries.ClassesDictionary;
import dictionaries.RelationshipsDictionary;
import util.CompilationResult;
import util.DataType;
import util.NamingManager;

import java.util.ArrayList;
import java.util.List;

public class CheckRelationship extends BaseOperator {

    /**
     * Конструктор
     * @param args Аргументы
     */
    public CheckRelationship(List<Operator> args) {
        super(args);
    }

    @Override
    public List<List<DataType>> argsDataTypes() {
        List<List<DataType>> result = new ArrayList<>();

        result.add(List.of(DataType.RELATIONSHIP, DataType.OBJECT));

        return result;
    }

    @Override
    public Boolean isArgsCountUnlimited() {
        return true;
    }

    @Override
    public DataType resultDataType() {
        return DataType.BOOLEAN;
    }

    @Override
    public CompilationResult compile() {
        // Объявляем переменные
        String value = "";
        String ruleHead = "";
        String completedRules = "";
        StringBuilder rulePartBuilder = new StringBuilder();
        StringBuilder completedRulesBuilder = new StringBuilder();

        // FIXME?: сейчас отношение можно получить только из value, если это изменится, тогда придется фиксить
        // Проверяем кол-во аргументов
        RelationshipValue relValue = (RelationshipValue) arg(0);
        String relName = relValue.value();
        if(RelationshipsDictionary.args(relName).size() != args().size() - 1) {
            throw new IllegalArgumentException("Некорректное количество аргументов");
        }

        // Компилируем аргументы
        List<CompilationResult> compiledArgs = new ArrayList<>();
        List<String> argValues = new ArrayList<>();
        for(Operator arg : args()) {
            compiledArgs.add(arg.compile());
            argValues.add(compiledArgs.get(compiledArgs.size()-1).value());
            rulePartBuilder.append(compiledArgs.get(compiledArgs.size() - 1).ruleHead());
            completedRulesBuilder.append(compiledArgs.get(compiledArgs.size() - 1).completedRules());
        }
        ruleHead = rulePartBuilder.toString();
        completedRules = completedRulesBuilder.toString();

        // ------------------- Добавляем переходы между классами --------------------

        for(int i = 1; i < args().size(); ++i) {
            String actualClass = "";
            String expectedClass = "";

            // ------------------- Определяем классы аргументов --------------------

            // TODO: добавить способы "Переменная, вводимая оператором", "Объект, полученный по условию"
            //  и "Экстремальный объект" после добавления таблицы
            // Если объект получен через отношение
            if(arg(i) instanceof GetByRelationship arg) {
                actualClass = RelationshipsDictionary.args(((RelationshipValue)arg.arg(1)).value()).get(1);
                expectedClass = RelationshipsDictionary.args(relName).get(i - 1);
            }
            // Если объект получен из переменной дерева
            else if (arg(i) instanceof DecisionTreeVarValue arg) {
                actualClass = ClassesDictionary.decisionTreeVarClass(arg.value());
                expectedClass = RelationshipsDictionary.args(relName).get(i - 1);
            }

            // Если классы не равны
            if(!expectedClass.equals(actualClass)) {
                // Если есть переход между классами
                if(ClassesDictionary.hasTransition(actualClass, expectedClass)) {
                    // TODO: определять необходимый объект в ситуации, когда связь не один к одному
                    String transitionRel = ClassesDictionary.transition(actualClass, expectedClass);
                    String transitionPattern = RelationshipsDictionary.pattern(transitionRel).first();

                    // FIXME?: добавить заполнение вспомогательных переменных, если появятся такие переходы
                    // Добавляем переход
                    String newArgValue = NamingManager.genVarName();

                    transitionPattern = transitionPattern.replace("<arg1>", argValues.get(i));
                    transitionPattern = transitionPattern.replace("<arg2>", newArgValue);
                    argValues.set(i, newArgValue);

                    ruleHead += transitionPattern;
                    completedRules += RelationshipsDictionary.pattern(transitionRel).second();
                }
                // Иначе - исключение
                else {
                    throw new IllegalArgumentException("Некорректный класс объекта " + i);
                }
            }
        }

        // ------------------- Добавляем проверку отношения --------------------

        String relPattern = RelationshipsDictionary.pattern(relName).first();
        for(int i = 1; i <= RelationshipsDictionary.args(relName).size(); ++i) {
            relPattern = relPattern.replace("<arg" + i + ">", argValues.get(i));
        }
        for(int i = 1; i <= RelationshipsDictionary.varCount(relName); ++i) {
            relPattern = relPattern.replace("<var" + i + ">", NamingManager.genVarName());
        }

        ruleHead += relPattern;
        completedRules += RelationshipsDictionary.pattern(relName).second();

        usedObjects = new ArrayList<>(argValues);
        usedObjects.remove(0);

        return new CompilationResult(value, ruleHead, completedRules);
    }
}

File: Compiler/src/compiler/operators/CheckRelationship.kt
package compiler.operators

import compiler.Literal
import compiler.Operator
import compiler.literals.RelationshipLiteral
import compiler.util.CompilationResult
import dictionaries.RelationshipsDictionary
import org.apache.jena.vocabulary.SchemaDO.pattern
import util.DataType
import util.NamingManager
import javax.xml.crypto.Data

/**
 * Проверка наличия отношения между объектами
 * TODO?: переходы между классами? relation type
 * TODO: negative form
 */
class CheckRelationship(args: List<Operator>) : BaseOperator(args) {

    /**
     * Является ли оператор негативным (т.е. нужно ли отрицание при компиляции)
     */
    internal var isNegative = false

    override val argsDataTypes
        get() = listOf(
            listOf(DataType.Relationship, DataType.Object)
        )


    override val isArgsCountUnlimited: Boolean
        get() = true

    override val resultDataType: DataType
        get() = DataType.Boolean


    override fun compile(): CompilationResult {
        // Объявляем переменные
        val heads = ArrayList<String>()
        var completedRules = ""
        val completedRulesBuilder = StringBuilder()

        // Компилируем аргументы
        val compiledArgs = ArrayList<CompilationResult>()
        val argValues = ArrayList<String>()

        for (arg in args) {
            val res = arg.compile()
            compiledArgs.add(res)
            argValues.add(res.value)
            completedRulesBuilder.append(res.rules)
        }

        completedRules += completedRulesBuilder

        // Если отношение вычисляется через другое отношение
        if (false) {
//            // Проверяем кол-во аргументов
//            val relValue = arg(0) as RelationshipLiteral
//            require(RelationshipsDictionary.args(relValue.value)!!.size == args.size - 2) {
//                "Некорректное количество аргументов"
//            }
//
//            // Получаем имя отношения
//            val mainRelName = (arg(0) as Literal).value
//            val supportRelName = (arg(1) as Literal).value
//
//            // Для всех результатов компиляции
//            val indices = ArrayList<Int>() // Индексы
//            compiledArgs.forEach { _ ->
//                indices.add(0)
//            }
//
//            // Пока не дошли до первого иднекса
//            while (indices.first() != compiledArgs.first().heads.size) {
//                // Собираем все части
//                var head = ""
//                for (i in compiledArgs.indices) {
//                    head += compiledArgs[i].heads[indices[i]]
//                }
//
//                // Добавляем проверку отношения
//                var pattern = pattern(mainRelName, supportRelName)!!.first
//                // Заполняем аргументы
//                for(i in RelationshipsDictionary.args(mainRelName)!!.indices) {
//                    pattern = pattern.replace("<arg$i>", argValues[i + 2])
//                }
//                // Заполняем переменные
//                for(i in 0 until varCount(mainRelName)!!) {
//                    pattern = pattern.replace("<var$i>", NamingManager.genVarName())
//                }
//
//                // Добавляем шаблон
//                head += pattern
//
//                // Добавляем в рзультат
//                heads.add(head)
//
//                // Меняем индексы
//                var i = indices.size - 1
//                while (true) {
//                    if (indices[i] != compiledArgs[i].heads.size - 1 || i == 0) {
//                        ++indices[i]
//                        break
//                    } else {
//                        indices[i] = 0
//                    }
//
//                    --i
//                }
//            }
//
//            // Сохраняем вспомогательные правила
//            completedRules += pattern(mainRelName)!!.second
        } else {
            // Проверяем кол-во аргументов
            val relValue = arg(0) as RelationshipLiteral
            require(RelationshipsDictionary.args(relValue.value)!!.size == args.size - 1) {
                "Некорректное количество аргументов"
            }

            // Получаем имя отношения
            val mainRelName = (arg(0) as Literal).value

            // Для всех результатов компиляции
            val indices = ArrayList<Int>() // Индексы
            compiledArgs.forEach { _ ->
                indices.add(0)
            }

            // Пока не дошли до первого иднекса
            while (indices.first() != compiledArgs.first().bodies.size) {
                // Собираем все части
                var head = ""
                for (i in compiledArgs.indices) {
                    head += compiledArgs[i].bodies[indices[i]]
                }

                // Добавляем проверку отношения
                var pattern =
                    if (isNegative) RelationshipsDictionary.negativeBody(mainRelName)!! else RelationshipsDictionary.body(
                        mainRelName
                    )!!
                // Заполняем аргументы

                for (i in RelationshipsDictionary.args(mainRelName)!!.indices) {

                    pattern = pattern.replace("<arg${i + 1}>", argValues[i + 1])
                }
                // Заполняем переменные
                for (i in 0..if (isNegative) RelationshipsDictionary.negativeVarsCount(mainRelName)!! else RelationshipsDictionary.varsCount(
                    mainRelName
                )!!) {
                    pattern = pattern.replace("<var$i>", NamingManager.genVarName())
                }

                // Добавляем шаблон
                head += pattern

                // Добавляем в рзультат
                heads.add(head)

                // Меняем индексы
                var i = indices.size - 1
                while (true) {
                    if (indices[i] != compiledArgs[i].bodies.size - 1 || i == 0) {
                        ++indices[i]
                        break
                    } else {
                        indices[i] = 0
                    }

                    --i
                }
            }

            // Сохраняем вспомогательные правила
            if (isNegative)
                RelationshipsDictionary.negativeRules(mainRelName)?.let {
                    completedRules += it
                }
            else
                RelationshipsDictionary.rules(mainRelName)?.let {
                    completedRules += it
                }

        }

        return CompilationResult("", heads, completedRules)
    }

    override fun clone(): Operator {
        val newArgs = ArrayList<Operator>()

        args.forEach { arg ->
            newArgs.add(arg.clone())
        }

        val res = CheckRelationship(newArgs)
        res.isNegative = isNegative
        return res
    }

    override fun clone(newArgs: List<Operator>): Operator {
        val res = CheckRelationship(newArgs)
        res.isNegative = isNegative
        return res
    }
}